feat(cost-model): migrate adv-stats

xx01cyx opened the pull request:
In `optd`, we use `MostCommonValues` and `Distribution` as trait bounds. Because of the serialization & deserialization need in `optd-persistent`, we use concrete types (`enum`) here instead of traits.

xx01cyx committed:
migrate stats

skyzh reviewed (approved):
I would rubber-stamp all patches in this repo to ensure we move fast but I'd like to note that

* Would be better to have an initial commit to copy-paste everything from the optd repo, and another commit which contains the changes to the original crate. Otherwise I don't know what changes I'll need to review :(
* Would be better if we can bring all commit history to this repo.

Looking at the dependencies, I saw lazy_static being used, which is already a deprecated crate and replaced by std libraries once cell. Probably needs to be fixed?

Regarding enum -- I'm not sure if this is a good approach regarding extensibility but let's merge it as-is.

xx01cyx commented:
> Would be better to have an initial commit to copy-paste everything from the optd repo, and another commit which contains the changes to the original crate. Otherwise I don't know what changes I'll need to review :(

Sure. This is actually separated from another larger PR. I just copy-pasted all code... I'll do so next time.

> Would be better if we can bring all commit history to this repo.

Is there a good way to do so?

> Looking at the dependencies, I saw lazy_static being used, which is already a deprecated crate and replaced by std libraries once cell. Probably needs to be fixed?

I've added a TODO in the code.

> Regarding enum -- I'm not sure if this is a good approach regarding extensibility but let's merge it as-is.

I've tried to make `MostCommonValues` and `Distribution` as traits, but doesn't work out very well. Not sure if there's a better solution.

skyzh commented:
https://git-scm.com/docs/git-filter-branch + git merge from an orphan branch
